libs/execution_context/ebpf_maps.c

```diff
@@ -525,100 +555,122 @@ _delete_per_cpu_hash_map(_In_ ebpf_core_map_t* map)
     per_cpu_data = (ebpf_core_per_cpu_data_t*)map->data;
     for (index = 0; index < per_cpu_data->count; index++) {
-        ebpf_hash_table_t** tables = (ebpf_hash_table_t**)&per_cpu_data->data;
-        ebpf_hash_table_destroy(tables[index]);
+        ebpf_core_per_cpu_hash_table_t* tables = (ebpf_core_per_cpu_hash_table_t*)&per_cpu_data->data;
+        ebpf_lock_destroy(&tables[index].lock);
+        ebpf_hash_table_destroy(tables[index].table);
+        tables[index].table = NULL;
     }
     ebpf_free(map);
 }
 
-static ebpf_hash_table_t*
-_get_hash_table_for_cpu(_In_ ebpf_core_map_t* map)
+static ebpf_core_per_cpu_hash_table_t*
+_get_hash_table_for_cpu(_In_ ebpf_core_map_t* map, uint32_t cpu_id)
 {
-    uint32_t current_cpu;
     ebpf_core_per_cpu_data_t* per_cpu_data = NULL;
-    ebpf_hash_table_t** tables;
-    if (ebpf_is_preemptible()) {
-        return NULL;
-    }
+    ebpf_core_per_cpu_hash_table_t* tables;
 
-    current_cpu = ebpf_get_current_cpu();
     per_cpu_data = (ebpf_core_per_cpu_data_t*)map->data;
-    tables = (ebpf_hash_table_t**)&per_cpu_data->data;
-    if (current_cpu < per_cpu_data->count) {
-        return tables[current_cpu];
+    tables = (ebpf_core_per_cpu_hash_table_t*)&per_cpu_data->data;
+    if (cpu_id < per_cpu_data->count) {
+        return tables + cpu_id;
     } else {
         return NULL;
     }
 }
 
 static uint8_t*
-_find_per_cpu_hash_map_entry(_In_ ebpf_core_map_t* map, _In_ const uint8_t* key)
+_find_per_cpu_hash_map_entry(_In_ ebpf_core_map_t* map, uint32_t cpu_id, _In_ const uint8_t* key)
 {
     uint8_t* value = NULL;
-    ebpf_hash_table_t* table;
+    ebpf_lock_state_t lock_state;
+    ebpf_core_per_cpu_hash_table_t* table;
     if (!map || !key)
         return NULL;
 
-    table = _get_hash_table_for_cpu(map);
+    table = _get_hash_table_for_cpu(map, cpu_id);
     if (!table)
         return NULL;
 
-    if (ebpf_hash_table_find(table, key, &value) != EBPF_SUCCESS) {
+    lock_state = ebpf_lock_lock(&table->lock);
+    if (ebpf_hash_table_find(table->table, key, &value) != EBPF_SUCCESS) {
         value = NULL;
     }
-
+    ebpf_lock_unlock(&table->lock, lock_state);
     return value;
 }
 
 static ebpf_result_t
-_update_per_cpu_hash_map_entry(_In_ ebpf_core_map_t* map, _In_ const uint8_t* key, _In_ const uint8_t* data)
+_update_per_cpu_hash_map_entry(
+    _In_ ebpf_core_map_t* map, uint32_t cpu_id, _In_ const uint8_t* key, _In_ const uint8_t* data)
 {
     ebpf_result_t result;
-    ebpf_hash_table_t* table;
+    ebpf_core_per_cpu_hash_table_t* table;
     size_t entry_count = 0;
     uint8_t* value;
-    if (!map || !key || !data)
+    ebpf_lock_state_t lock_state;
+
+    if (!map || !key)
         return EBPF_INVALID_ARGUMENT;
 
-    table = _get_hash_table_for_cpu(map);
+    table = _get_hash_table_for_cpu(map, cpu_id);
     if (!table)
         return EBPF_INVALID_ARGUMENT;
 
-    entry_count = ebpf_hash_table_key_count(table);
+    lock_state = ebpf_lock_lock(&table->lock);
```

Comment:
This is even worse for perf since you take the lock once inside 615, release it, and then take it again.  Every lock acquisition/release is expensive.  We should strive to make this have 0 such transitions, rather than 4.

Reply:
Thanks. Updated the logic to update map entries.

The fast path is:
1) Invoked from the helper function.
2) An entry for this key is already present.

If 1 & 2 are met, then update the existing map entry with the new value.

The slow path is:
1) Insert zero entries into each per-cpu map if not present.
2) Update either current per-CPU entry or all per-CPU entries (from helper vs from caller).

The slow path is going to be expensive no matter what we do as it will need to acquire O(N) locks where N is count of CPUs.